### ISSUE TITLE
mongo oplog_observe_driver "tried to add something already published" error thrown

### ISSUE THREAD

**Issue by conorstrejcek** [meteor/meteor#10512]
## Version Information
Meteor version: 1.8.1-rc.1
Last working version: N/A
OS: Mac OS

## Expected Behavior
Oplog should work without errors when using subscriptions/publications with transactions.

##  Actual Behavior
While it looks like #10420 fixes one of the bugs with the oplog driver, there is a new error which now appears in our logs (and breaks reactivity):

```
Got exception while fetching documents Error: tried to add something already published 9jXBDutmDgrvTJ8yC
      at packages/mongo/oplog_observe_driver.js:330:15
      at Object.Meteor._noYieldsAllowed (packages/meteor.js:778:12)
      at OplogObserveDriver._addMatching (packages/mongo/oplog_observe_driver.js:327:12)
      at packages/mongo/oplog_observe_driver.js:396:14
      at Object.Meteor._noYieldsAllowed (packages/meteor.js:778:12)
      at OplogObserveDriver._handleDoc (packages/mongo/oplog_observe_driver.js:388:12)
      at packages/mongo/oplog_observe_driver.js:529:26
      at packages/mongo/oplog_observe_driver.js:15:9
      at packages/mongo/doc_fetcher.js:49:11
```

## Reproduction
I am having difficulty reproducing this, as it only seems to happen when there are many concurrent users. I was able to determine that this error is most likely thrown in response to a successful transaction which would cause a new document to match in an existing publication. Beyond that, I am having some trouble causing the issue locally.

If it is not possible for someone to look into this with the limited information I can provide, does anyone have a suggestion for how I can easily implement local changes to the `oplog_observe_driver` which persist across builds (so that I can deploy to our production instance with some additional logging)?

EDIT: I have put together a reproduction here: https://github.com/conorstrejcek/oplog-error-repro/tree/master

**Comment by conorstrejcek** [comment]
I was finally able to reproduce the issue locally: https://github.com/conorstrejcek/oplog-error-repro/tree/master

There is a bit of complex logic involved in generating the operations for the transaction, but you can ignore that and just look at the actual updates which were run, as well as the logic of the publication itself.

**Comment by KoenLav** [comment]
We are running into the same issue on one of our staging environments, don't know which portion of code causes this yet.

**Comment by chrisbobbe** [comment]
@benjamn Is this issue in a good state to mark with some flags, like "bug" and "has-reproduction," etc.? It affects anyone using oplog tailing with transactions.

**Comment by benjamn** [comment]
@chrisbobbe @conorstrejcek @KoenLav Sorry for not getting to this sooner! I can confirm that my commit b0cb1ceda3dbe7e1b505d0017f2ccc50d6eb7c40 makes the problem go away in @conorstrejcek's reproduction, though I have to admit I'm concerned about the general strategy of using `Timestamp`s to identify operations. I'll keep thinking about this, but in the meantime please try the next beta of Meteor 1.8.2 (likely `1.8.2-beta.7`).

**Comment by chrisbobbe** [comment]
Thanks very much, I’ll check out the next beta tomorrow! If it helps with
your concern (and I’m not sure it does), the Timestamp type is different
from the Date type, in that it’s guaranteed to be unique within a single
mongod instance; it includes an ordinal number when there are multiple
records in one second.

**Comment by benjamn** [comment]
Closing because I believe this issue was fixed as of `1.8.2-beta.12`, though please reopen if the problem is still present.
